L82-core-event-engine.md

```diff
@@ -0,0 +1,538 @@
+
+### Related Proposals:
+
+* This work supersedes [L69: EventManger API](https://github.com/grpc/proposal/pull/182)
+* Supports [L67: C++ callback-based asynchronous API](https://github.com/grpc/proposal/pull/180) as a more performant solution than provided by the [CallbackAlternativeCQ](https://github.com/grpc/grpc/pull/25169).
+
+## Proposal
+
+### Concepts
+
+An EventEngine is roughly composed of:
+
+* A task runner which asynchronously executes callables with optional
+  scheduling.
+* An asynchronous DNS resolver.
+* Network listener logic that accepts incoming connection requests, and
+  performs network-server-related I/O operations.
+* Client logic, support for making outgoing network connections.
+
+### Common Data Types
+
+```cpp
+namespace grpc_event_engine {
+namespace experimental {
+
+/// A set of parameters used to configure an endpoint, either when initiating a
+/// new connection on the client side or when listening for incoming connections
+/// on the server side. An EndpointConfig contains a set of zero or more
+/// Settings. Each setting has a unique name, which can be used to fetch that
+/// Setting via the Get() method. Each Setting has a value, which can be an
+/// integer, string, or void pointer. Each EE impl should define the set of
+/// Settings that it supports being passed into it, along with the corresponding
+/// type.
+class EndpointConfig {
+ public:
+  virtual ~EndpointConfig() = default;
+  using Setting = absl::variant<absl::monostate, int, absl::string_view, void*>;
+  /// Returns an EndpointConfig Setting. If there is no Setting associated with
+  /// \a key in the EndpointConfig, an \a absl::monostate type will be
+  /// returned. Caller does not take ownership of resulting value.
+  virtual Setting Get(absl::string_view key) const = 0;
+};
+```
+
+The `SliceAllocator` provides memory buffers for EventEngine read/write
+operations, allowing EventEngines to seamlessly participate in the
+C++ `ResourceQuota` system.
+
+```cpp
+class SliceAllocator {
+ public:
+  virtual ~SliceAllocator() = default;
+  using AllocateCallback = std::function<void(absl::Status)>;
+  /// Requests \a size bytes from gRPC, and populates \a dest with the allocated
+  /// slices. Ownership of the \a SliceBuffer is not transferred.
+  ///
+  /// gRPC provides a ResourceQuota system to cap the amount of memory used by
+  /// the library. When a memory limit has been reached, slice allocation is
+  /// interrupted to attempt to reclaim memory from participating gRPC
+  /// internals. When there is sufficient memory available, slice allocation
+  /// proceeds as normal.
+  virtual absl::Status Allocate(size_t size, SliceBuffer* dest,
+                        SliceAllocator::AllocateCallback cb) = 0;
+};
+```
+
+The `SliceAllocatorFactory` allows `EventEnigine::Listener`s to create and
+provide `SliceAllocator`s to `EventEngine::Endpoint`s as incoming connections
+are accepted.
+
+```cpp
+class SliceAllocatorFactory {
```

Comment:
What is the benefit of having a `SiceAllocator` and a `SliceAllocatorFactory`?
Are there expected locality or lifecycle conciderations?
I would recommend to move the `peer_name` to Allocate and to remove the factory.

Reply:
As explained above, the SliceAllocator transparently enrolls EventEngines into the ResourceQuota system. The ResourceQuota system is a public feature, so we can't break that. Therefore something like the SliceAllocator needs to exist. The SliceAllocatorFactory mirrors the ResourceQuota itself (which is a parent to multiple independent resource users that can be reclaimed separately, incrementally if the quota has been exhausted), while the SliceAllocator hides the resource users within. I believe Craig has started some public documentation on the resource quota feature, it's fairly sparse at the moment.

The lifetimes are managed by smart pointers in the Connect and CreateListener APIs here. I believe our main implementations will let these objects remain valid until gRPC is shut down. Do you have lifetime questions that aren't answered by that?

The `peer_name` is an identifier for the Allocator (for the endpoint, essentially), not for a single allocation request, so it should remain in the constructor.